### DISCUSSION TITLE
Create user failing in keycloak.

### DISCUSSION THREAD

**Discussion by ytk1389** [Sunbird-Lern/Community#100] [Issues]
Hi team,
We have integrated user-org service with the keycloak server and trying to create a user using the user create api in it.When I am trying to create the user i am getting the below exception "**Method threw 'javax.ws.rs.NotFoundException' exception.
HTTP 404 Not Found**" ,when I am hitting the same url via  webbrowser i am able to see the keycloak console ((http://localhost:8080/auth/)(which shows keycloak server is running successfully) .
Could someone please help me resolve this issue.
Note:Basically we trying to create a new user and later will also try to retrieve it.

**Comment by BharathwajShankar** [comment]
@ytk1389 please share the details of the API, request and response you tried along with the exact stack trace you got in the console. It will help us to suggest you better steps.

**Comment by ytk1389** [comment]
**API** : {user-org-service-url}/v1/user/create
**Request** : {
  "request": {
    "firstName": "Test29",
    "userName": "Test29",
    "email": "Test29@mailinator.com",
    "lastName": "Test28",
    "emailVerified": true,
    "password": "Test28@1238",
    "phone":"1234561128",
    "countrycode":"+91"
  }
}
**Response** : {
    "id": "api.user.create",
    "ver": "v1",
    "ts": "2023-04-08 16:11:19:200+0530",
    "params": {
        "resmsgid": "f3fd3770-5246-4317-83f2-da71f1a870c7",
        "msgid": "f3fd3770-5246-4317-83f2-da71f1a870c7",
        "err": "UOS_USRCRT0080",
        "status": "FAILED",
        "errmsg": "SERVICE UNAVAILABLE"
    },
    "responseCode": "OK",
    "result": {}
}


\"stacktrace\":\" Exception Occured  org.jboss.resteasy.client.jaxrs.engines.ApacheHttpClient4Engine.invoke(ApacheHttpClient4Engine.java:289)org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:454)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:105)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)com.sun.proxy.$Proxy54.grantToken(Unknown Source)org.keycloak.admin.client.token.TokenManager.grantToken(TokenManager.java:89)org.keycloak.admin.client.token.TokenManager.getAccessToken(TokenManager.java:69)org.keycloak.admin.client.token.TokenManager.getAccessTokenString(TokenManager.java:64)org.keycloak.admin.client.resource.BearerAuthFilter.filter(BearerAuthFilter.java:52)org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.invoke(ClientInvocation.java:431)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:105)org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)com.sun.proxy.$Proxy69.create(Unknown Source)org.sunbird.sso.impl.KeyCloakServiceImpl.createUser(KeyCloakServiceImpl.java:281)org.sunbird.util.user.UserUtil.addUser(UserUtil.java:307)org.sunbird.service.user.impl.SSOUserServiceImpl.createUserAndPassword(SSOUserServiceImpl.java:99)org.sunbird.actor.user.SSOUserCreateActor.processSSOUser(SSOUserCreateActor.java:123)org.sunbird.actor.user.SSOUserCreateActor.createSSOUser(SSOUserCreateActor.java:89)org.sunbird.actor.user.SSOUserCreateActor.onReceive(SSOUserCreateActor.java:61)org.sunbird.actor.core.BaseActor.onReceive(BaseActor.java:21)akka.actor.UntypedAbstractActor$$anonfun$receive$1.applyOrElse(AbstractActor.scala:291)akka.actor.Actor$class.aroundReceive$$$capture(Actor.scala:517)akka.actor.Actor$class.aroundReceive(Actor.scala)akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:180)akka.actor.ActorCell.receiveMessage$$$capture(ActorCell.scala:588)akka.actor.ActorCell.receiveMessage(ActorCell.scala)akka.actor.ActorCell.invoke(ActorCell.scala:557)akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)akka.dispatch.Mailbox.run(Mailbox.scala:225)akka.dispatch.Mailbox.exec(Mailbox.scala:235)akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\",\"errtype\":\"api_access\",\"requestid\":\"092c32ab-fddd-4445-a8fa-c906b316dc2c\"}}",
**NOTE:Though the user is getting persisted in cassandra database  and the entry is visible in database,facing the above issue when it comes to keycloak**
Your help to overcome this issue will be highly appreciated.
Thanks.

**Reply by reshmi-nair** [reply]
@ytk1389 Pls check the keycloak configurations. The error clearly says "SERVICE UNAVAILABLE".  Can you check the keycloak related env variables like "sunbird_keycloak_user_federation_provider_id", sunbird_sso_url, sunbird_sso_realm, sunbird_sso_client_secret, sunbird_sso_client_id etc.

**Reply by ytk1389** [reply]
Thanks for your reply,we went through that and currently facing the below issue
{
    "id": "api.user.create",
    "ver": "v1",
    "ts": "2023-04-11 10:59:46:109+0530",
    "params": {
        "resmsgid": "6fbac99e-c963-465d-b586-0db4ec9aa85d",
        "msgid": "6fbac99e-c963-465d-b586-0db4ec9aa85d",
        "err": "UOS_USRCRT0080",
        "status": "FAILED",
        "errmsg": "Keycloak Server Error"
    },
    "responseCode": "OK",
    "result": {}
}
Below is the stacktrace:
	"msg": "{\"eid\":\"ERROR\",\"ets\":1681190986109,\"ver\":\"3.0\",\"mid\":\"6fbac99e-c963-465d-b586-0db4ec9aa85d\",\"actor\":{\"id\":\"ccc1f6ac-2461-4769-8d21-7930f206d434\",\"type\":\"User\"},\"context\":{\"channel\":\"ORG_001\",\"pdata\":{\"id\":\"local.sunbird.learning.service\",\"pid\":\"learning-service\",\"ver\":\"5.0.0\"},\"env\":\"User\",\"cdata\":[{\"id\":\"6fbac99e-c963-465d-b586-0db4ec9aa85d\",\"type\":\"Request\"}],\"rollup\":{}},\"edata\":{\"err\":\"UOS_USRCRT0080\",\"stacktrace\":\"Keycloak Server Error org.sunbird.sso.impl.KeyCloakServiceImpl.createUser(KeyCloakServiceImpl.java:286)org.sunbird.util.user.UserUtil.addUser(UserUtil.java:307)org.sunbird.service.user.impl.SSOUserServiceImpl.createUserAndPassword(SSOUserServiceImpl.java:99)org.sunbird.actor.user.SSOUserCreateActor.processSSOUser(SSOUserCreateActor.java:124)org.sunbird.actor.user.SSOUserCreateActor.createSSOUser(SSOUserCreateActor.java:89)org.sunbird.actor.user.SSOUserCreateActor.onReceive(SSOUserCreateActor.java:61)org.sunbird.actor.core.BaseActor.onReceive(BaseActor.java:21)akka.actor.UntypedAbstractActor$$anonfun$receive$1.applyOrElse(AbstractActor.scala:291)akka.actor.Actor$class.aroundReceive$$$capture(Actor.scala:517)akka.actor.Actor$class.aroundReceive(Actor.scala)akka.actor.AbstractActor.aroundReceive(AbstractActor.scala:180)akka.actor.ActorCell.receiveMessage$$$capture(ActorCell.scala:588)akka.actor.ActorCell.receiveMessage(ActorCell.scala)akka.actor.ActorCell.invoke(ActorCell.scala:557)akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)akka.dispatch.Mailbox.run(Mailbox.scala:225)akka.dispatch.Mailbox.exec(Mailbox.scala:235)akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)\",\"errtype\":\"api_access\",\"requestid\":\"6fbac99e-c963-465d-b586-0db4ec9aa85d\"}}",
Internally i am getting **javax.ws.rs.NotFoundException: HTTP 404 Not Found** exception
![Screenshot from 2023-04-11 11-25-51](https://user-images.githubusercontent.com/50234913/231069075-24358ebc-390b-42e5-b14d-17aa38d099b5.png)
Your guidance through this will be highly appreciated.
Thanks in advance

**Reply by Hari-stackroute** [reply]
@ytk1389  all the environment variables are properly configured with respect to keycloak, schedule a call if possible.

**Reply by ytk1389** [reply]
Thanks for your reply,could you please let me know how i can reach out to you and get on call?

**Reply by Hari-stackroute** [reply]
schedule call in google meet and share us the link. @ytk1389

**Reply by ytk1389** [reply]
https://meet.google.com/hqm-bgom-rgj?authuser=1&hl=en_GB , we can connect on this link,please let me know when you shall be available for us to connect @Hari-stackroute .
Thanks

**Reply by Hari-stackroute** [reply]
@ytk1389 we can connect at 2:15 pm

**Reply by ytk1389** [reply]
> @ytk1389 we can connect at 2:15 pm

on the bridge https://meet.google.com/hqm-bgom-rgj?authuser=1&hl=en_GB

**Reply by Jayaprakash8887** [reply]
@ytk1389 Please confirm if the issue is resolved. If yes, kindly mark the ticket as answered. Kindly also mention steps taken to resolve the issue

**Reply by Jayaprakash8887** [reply]
@ytk1389 Please update

**Comment by AmiableAnil** [comment]
Closing this as no update.
